### ISSUE TITLE
ATM/Bank System

### ISSUE THREAD

**Issue by PatchRequest** [EveronLife/EveronLife#10]
**Summarize your proposal**
A system for players to interact with an object or NPC to transfer money from/to their account and to other players

**Describe the solution you'd like**
The system should provide a GUI for players to 
- Deposit money in their account
- Withdraw money from their account
- send money to other players

Potential fees should be implemented as a parameter
The implemented interfaces of the existing Money System should be used

In the future, there could be a difference between a bank NPC and an ATM.
In the future, there could be different types of accounts to store money (and get interest).

**Comment by Kabun402** [comment]
I would like to contribute to this topic with a quite big feature set so I'm asking if this is suitable for a public framework or already over the top befor I waste the time that I would spend on it.

My plans for an Bank/Money Systems would be as follows:
- an bank account is not literaly connected to an player but to an bank card (ownsership of an bank account will still be a player maybe also a faction?)
- a player can own N bank accounts
- the owner of an bank account is permitted to create N bank cards and give them to other players
- the owner can view a list of all cards for his bank account and set a comment for each to for example hold track of who has which card
- to access a bank account at an ATM you need the card and the pin for this card which consists of N characters/numbers
- if a pin is entered wrong N times the card is locked and kept by the ATM
- the owner can lock a card so it can't be used anymore and will be kept by the ATM the next time it is inserted
- the owner of the bank account can set limits for every card for how much is allowed to deposit, withdraw and transfered per timespan
- police or other law institutions can view history, freeze an account or force a transfer (with check for rank)
- there are mulitple bank institutions available to choose from with different conditions
   - fee for creating an account
   - fee for creating an card
   - fee for deposit, withdraw and transfer money
   - maybe interest for a defined timespan
   - police or other law institutions will see only every n-th history entry (or something similar for illegal accounts)
- there will exist an system bank account so any transfer that would come from or go to the "void" (like trading with NPC's, paycheck for police, medic or whatever) instead goes to/comes from the system bank account
therefor the NPC's would have to have a card linked to the system bank account (all NPS have the "same" card or every NPS gets his own) this allows administrators (or maybe even police or other law institutions) to monitor money flow

I would do my best to implement the whole thing as configurable as possible so server owners can decide which features they want and which not but obviusly there are limitations as some features would have dependencies of others.

Also I'm absolutly no frontend guy so I would need help or someone that completly does the GUI.

Opinions and other suggestions for this system are welcome.

**Comment by MiniFailPro** [comment]
What if u lose your card? How to get a new one?

**Comment by Kabun402** [comment]
> What if u lose your card? How to get a new one?

"the owner of an bank account is permitted to create N bank cards"
so the owner would have to go (propably) to an NPC of the given bank institution an get a new one

**Comment by jackdevonshire** [comment]
I think this is a nice idea, but there would be a whole host of issues to resolve. The bank card would take up another inventory slot as cash is also being discussed as a physical item. Also, if a player gets robbed, I'm not particularly fond of the idea that a rebel could wipe their bank account and all of their monetary progress clean, in a matter of seconds - and if a mechanism was added to lock a cards usage to a specific player, it just removes the need for them all together right?

Final comment from me, for realism it is nice, but does this not just add another boring chore that players have to do to access their in game resources? I remember in Altis Life, every time you died you would have to go to an ATM to get money out, take a car out, drive to get clothes, drive to get virtual items, drive to restock physical items, drive to get rebel clothes etc - sure all of these are actually essential, but I worry that adding any more to this flow goes from "making it more real with a cool system" to just another pointless chore that players probably don't want to do...Especially in the core framework.

Didn't mean to ... all over the idea, I think some sort of fancy banking mechanic would be nice, just don't like the idea of physical bank cards.

**Comment by Jaccodouma** [comment]
> but does this not just add another boring chore that players have to do to access their in game resources

I think this is huge. I reckon just having your account and being able to do transactions anywhere from there makes more sense for most stuff. Kabun's idea seems cool, but probably too compilated and I'm not too sure how much it adds over simpler implementations. (obv also just my opinion)
If a player just has an "account" (probably doesn't need to be more than a money var) which they can use to make purchases and transfer money initially we should have enough right? Maybe track transactions for police to look at. Physical money would be cool for illegal transactions, ie fill a briefcase with money to buy weapons from some smuggler or something, but I dont think it should be the basis of all transactions. Always disliked that in A3 life too.

**Comment by Kabun402** [comment]
I'm aware that some server owners might not want to have all the features that I listed.
As I said I would implement the whole system as configurable as possible which means you can decide for example:
 - if you want the cards to be an actual item or some kind of virtual item or even if you want the whole card system at all
 - if you have cards enabled should the ATM prompt for a pin
 - should the be an system bank account
by editing a config file. (I think there it comes in handy that I almost always write code very generic and configurable.)

So if a server owner wants he can configure the system in such a way that it works like in altis life he can but if he wants to have a system that is as realistic as possible he can too.
I certainly know that this isn't exactly easy and fast to implement but I'm willing to try my best.

**Comment by DrLuk** [comment]
I like the more complex process as an option you could configure. But I think the default should be a simple and easy system, like proposed at the beginning.
Something I would reccomend adding is the option to pay for stuff without having to withdraw the cash. If you want to balance it you can still put a fee on the "virtual" transaction.

**Comment by ebork11** [comment]
> Something I would reccomend adding is the option to pay for stuff without having to withdraw the cash. If you want to balance it you can still put a fee on the "virtual" transaction.

This could be treated as a sort of "debit card" sort of option. A player could purchase a debit card from an ATM or banker and this would enable them to make purchases at vendors without having the actual cash on them, however it would come with a fee associated with it. Additionally, there could be different "tiers" of cards that have different spending limits so that this could not be abused. For example, tier 1 has $10,000 able to be spent in the next hour. 

By doing this it does remove the ability to rob someone for their cash as they travel around and purchase things, completely removing that risk. Something could be added allowing the robber to attempt to steal a partial amount from the persons bank account through either pin guessing, selling it, hacking it, etc.. This amount could be based on the tier mentioned above so that it never bankrupts an entire person and its also not a guarantee, providing them some protection.

**Comment by KasperOnPC** [comment]
Just concerning the **basic system** looking at the existing assets I had the following Ideas.

**Depositing Money**
These secured post boxes could be used to send money but not withdraw it. This would allow players to easily "post" a money deposit to the bank. Maybe the money even sits in it for 5 minutes before it is "sent" meaning that gangs could potentially rob mailboxes. 
![image](https://user-images.githubusercontent.com/106525617/172033389-eb0a51f5-0933-42ef-a679-0d109bac78f5.png)

**Depositing and Withdrawing Money**
The below municipal building has a lower floor with a very bank-like room with a lobby/waiting room and offices upstairs, much like a standard retail bank in real life. One office even looks like it could be the bank manager's office. 
![image](https://user-images.githubusercontent.com/106525617/172033407-dd027bef-4401-4138-a53b-b2da370562a0.png)
**Banking Room**
Here's what the bank-like area looks like where players could deposit and withdraw money, I feel that this would be safer than the standard ATMs you had in Altis but they would also act as another potential location for gangs to rob. With multiple levels and doors and its close proximity to the police buildings, it would make it a great location to keep players safe and provide a really difficult heist for anyone willing to try it. 
![image](https://user-images.githubusercontent.com/106525617/172033453-4a792e15-ec50-4043-825a-872790cd9abe.png)

Keen for any thoughts, I'm not sure about how the "storage" of money will work, I assume in the SQL server but let me know how I can contribute.

**Comment by Friiize** [comment]
That's nice. Also, i don't know if the RP is set at a specific era but if we can for exemple, use mobile phone for transactions, we should considere that part too instead of mail boxes ? I didn't even know it was a thing before reading your post. 

In my opinion, the ATMs should remains, with perhaps some tweaks that make some interactions with your money only possible here and not on the smartphone, like for exemple a large amount of money transfer. And since i mention the mobile phone, is it interesting to implement hacking phone or any electronics ? Too much ?

**Comment by DrLuk** [comment]
I like your creative ideas. My concern with money related things is just that it can get frustrating. 
Once you got paid after farming clay for like 2h on a server, you should be able to hold onto it and use it somewhat safely without the possibility of getting robbed 5 times between your paycheck and buying a new car.
Money is the base for all the other features, if you constantly need to run around depositing and withdrawing money from different places and watch out for criminals you won't ever have time for other things. 
Overdoing the banking system is easy, but I fear people could get frustrated pretty fast when the experience is more a pain than fun.

**Comment by jackdevonshire** [comment]
Going to jump back on this, I'm with DrLuk. I think the money system should be as simple as possible with some hint of realism, mostly mirroring the system in Altis Life - I would argue cash shouldn't be required at every store however. There are plenty of other systems that can be overcomplicated for the sake of realism - so that 'succeeding' in Everon Life isn't super easy - but I don't think the money system is something that should be played with, because people will get bored of it extremely fast.

**Comment by KasperOnPC** [comment]
> Money is the base for all the other features, if you constantly need to run around depositing and withdrawing money from different places and watch out for criminals you won't ever have time for other things. 

This is true, I think this could be solved by simply adding alot of the deposit boxes could fix that. We could also make it so you could withdraw at them if we wanted to make it 1:1 with Altis Life.

**Comment by DrLuk** [comment]
Yes. I think at least for the beginning we could just go for those boxes or just plain ATMs. Plus paying certain things wihtout cash, by directly deducting from the bank account, as I mentioned before.

**Comment by njbrown09** [comment]
I think ATMS where you just deposit and withdrawl cash are the best idea. Simple is the best way to go here.

**Comment by DrLuk** [comment]
I have an ATM model that I never got to finish, we could use that. It just needs some textures.
![grafik](https://user-images.githubusercontent.com/16528763/172219761-07ab005e-2ddb-4346-8218-c61fa4a4896b.png)

Or we can go a less modern style, maybe also both for variation.

**Comment by KasperOnPC** [comment]
This one also looks pretty good:
https://sketchfab.com/3d-models/atm-17480b0cc8364184972f126513924372

I just learnt that ATMs are actually in line with the map's time period, so that's cool.

**Comment by sufiyanmirza6454-art** [comment]
I want to learn hacking.
